File: atlhyper_web/src/api/workload.ts
/**
 * Workload API (StatefulSet & DaemonSet)
 *
 * 后端已完成 model_v2 → model 扁平化转换，前端直接使用
 */

import { get } from "./request";

// ============================================================
// 嵌套子结构类型（匹配 model_v2 JSON 结构）
// ============================================================

interface WorkloadRollout {
  phase: string;
  message?: string;
  badges?: string[];
}

interface WorkloadCondition {
  type: string;
  status: string;
  reason?: string;
  message?: string;
  lastUpdateTime?: string;
  lastTransitionTime?: string;
}

interface UpdateStrategy {
  type?: string;
  partition?: number;
  maxUnavailable?: string;
  maxSurge?: string;
}

interface ContainerPort {
  name?: string;
  containerPort: number;
  protocol?: string;
}

interface Probe {
  type: string;
  path?: string;
  port?: number;
  command?: string;
}

interface ContainerDetail {
  name: string;
  image: string;
  imagePullPolicy?: string;
  ports?: ContainerPort[];
  requests?: Record<string, string>;
  limits?: Record<string, string>;
  livenessProbe?: Probe;
  readinessProbe?: Probe;
  startupProbe?: Probe;
}

interface VolumeSpec {
  name: string;
  type: string;
  source?: string;
}

interface Toleration {
  key?: string;
  operator?: string;
  value?: string;
  effect?: string;
}

interface PodTemplate {
  containers: ContainerDetail[];
  volumes?: VolumeSpec[];
  nodeSelector?: Record<string, string>;
  tolerations?: Toleration[];
}

// StatefulSet 专用

interface PVCRetentionPolicy {
  whenDeleted?: string;
  whenScaled?: string;
}

interface VolumeClaimTemplate {
  name: string;
  accessModes?: string[];
  storageClass?: string;
  storage?: string;
}

interface StatefulSetSpec {
  podManagementPolicy?: string;
  updateStrategy?: UpdateStrategy;
  minReadySeconds?: number;
  revisionHistoryLimit?: number;
  persistentVolumeClaimRetentionPolicy?: PVCRetentionPolicy;
  volumeClaimTemplates?: VolumeClaimTemplate[];
}

interface StatefulSetStatus {
  currentRevision?: string;
  updateRevision?: string;
}

// DaemonSet 专用

interface DaemonSetSpec {
  updateStrategy?: UpdateStrategy;
  minReadySeconds?: number;
  revisionHistoryLimit?: number;
}

// ============================================================
// 导出类型（供组件使用）
// ============================================================

export interface StatefulSetDetail {
  // 扁平顶层（从 summary 提取）
  name: string;
  namespace: string;
  replicas: number;
  ready: number;
  current: number;
  updated: number;
  available: number;
  createdAt: string;
  age: string;
  serviceName?: string;
  selector?: string;

  // 嵌套子结构（后端透传 model_v2 原始结构）
  spec: StatefulSetSpec;
  template?: PodTemplate;
  status: StatefulSetStatus;
  conditions?: WorkloadCondition[];
  rollout?: WorkloadRollout;

  // 元数据
  labels?: Record<string, string>;
  annotations?: Record<string, string>;
}

export interface DaemonSetDetail {
  // 扁平顶层（从 summary 提取，字段已重命名）
  name: string;
  namespace: string;
  desired: number;
  current: number;
  ready: number;
  available: number;
  unavailable: number;
  misscheduled: number;
  updatedScheduled: number;
  createdAt: string;
  age: string;
  selector?: string;

  // 嵌套子结构（后端透传 model_v2 原始结构）
  spec: DaemonSetSpec;
  template?: PodTemplate;
  conditions?: WorkloadCondition[];
  rollout?: WorkloadRollout;

  // 元数据
  labels?: Record<string, string>;
  annotations?: Record<string, string>;
}

// ============================================================
// 响应类型
// ============================================================

interface ListParams {
  cluster_id: string;
  namespace?: string;
}

// StatefulSet list still returns raw model_v2 (no list item conversion)
interface StatefulSetListResponse {
  message: string;
  data: unknown[];
  total: number;
}

interface DaemonSetListResponse {
  message: string;
  data: unknown[];
  total: number;
}

// ============================================================
// StatefulSet API
// ============================================================

/**
 * 获取 StatefulSet 列表
 */
export function getStatefulSetList(params: ListParams) {
  return get<StatefulSetListResponse>("/api/v2/statefulsets", params);
}

/**
 * 获取 StatefulSet 详情
 */
export async function getStatefulSetDetail(data: { ClusterID: string; Namespace: string; Name: string }) {
  return get<{ message: string; data: StatefulSetDetail }>(
    `/api/v2/statefulsets/${encodeURIComponent(data.Name)}`,
    {
      cluster_id: data.ClusterID,
      namespace: data.Namespace,
    }
  );
}

// ============================================================
// DaemonSet API
// ============================================================

/**
 * 获取 DaemonSet 列表
 */
export function getDaemonSetList(params: ListParams) {
  return get<DaemonSetListResponse>("/api/v2/daemonsets", params);
}

/**
 * 获取 DaemonSet 详情
 */
export async function getDaemonSetDetail(data: { ClusterID: string; Namespace: string; Name: string }) {
  return get<{ message: string; data: DaemonSetDetail }>(
    `/api/v2/daemonsets/${encodeURIComponent(data.Name)}`,
    {
      cluster_id: data.ClusterID,
      namespace: data.Namespace,
    }
  );
}

File: model_v2/workload.go
package model_v2

import "time"

// ============================================================
// StatefulSet 模型（嵌套结构）
// ============================================================

// StatefulSet K8s StatefulSet 资源模型
type StatefulSet struct {
	Summary  StatefulSetSummary  `json:"summary"`
	Spec     StatefulSetSpec     `json:"spec"`
	Template PodTemplate         `json:"template"`
	Status   StatefulSetStatus   `json:"status"`
	Rollout  *WorkloadRollout    `json:"rollout,omitempty"`

	// 元数据
	Labels      map[string]string `json:"labels,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`
}

// StatefulSetSummary StatefulSet 摘要
type StatefulSetSummary struct {
	Name        string    `json:"name"`
	Namespace   string    `json:"namespace"`
	Replicas    int32     `json:"replicas"`
	Ready       int32     `json:"ready"`
	Current     int32     `json:"current"`
	Updated     int32     `json:"updated"`
	Available   int32     `json:"available"`
	CreatedAt   time.Time `json:"createdAt"`
	Age         string    `json:"age"`
	ServiceName string    `json:"serviceName,omitempty"`
	Selector    string    `json:"selector,omitempty"`
}

// StatefulSetSpec StatefulSet 规格
type StatefulSetSpec struct {
	Replicas             *int32               `json:"replicas,omitempty"`
	ServiceName          string               `json:"serviceName,omitempty"`
	PodManagementPolicy  string               `json:"podManagementPolicy,omitempty"` // OrderedReady, Parallel
	UpdateStrategy       *UpdateStrategy      `json:"updateStrategy,omitempty"`
	RevisionHistoryLimit *int32               `json:"revisionHistoryLimit,omitempty"`
	MinReadySeconds      int32                `json:"minReadySeconds,omitempty"`
	PersistentVolumeClaimRetentionPolicy *PVCRetentionPolicy `json:"persistentVolumeClaimRetentionPolicy,omitempty"`
	Selector             *LabelSelector       `json:"selector,omitempty"`
	VolumeClaimTemplates []VolumeClaimTemplate `json:"volumeClaimTemplates,omitempty"`
}

// UpdateStrategy 更新策略
type UpdateStrategy struct {
	Type          string `json:"type,omitempty"` // RollingUpdate, OnDelete
	Partition     *int32 `json:"partition,omitempty"`
	MaxUnavailable string `json:"maxUnavailable,omitempty"`
	MaxSurge       string `json:"maxSurge,omitempty"`
}

// PVCRetentionPolicy PVC 保留策略
type PVCRetentionPolicy struct {
	WhenDeleted string `json:"whenDeleted,omitempty"` // Retain, Delete
	WhenScaled  string `json:"whenScaled,omitempty"`  // Retain, Delete
}

// VolumeClaimTemplate PVC 模板
type VolumeClaimTemplate struct {
	Name         string            `json:"name"`
	AccessModes  []string          `json:"accessModes,omitempty"`
	StorageClass string            `json:"storageClass,omitempty"`
	Storage      string            `json:"storage,omitempty"`
	Selector     map[string]string `json:"selector,omitempty"`
}

// StatefulSetStatus StatefulSet 状态
type StatefulSetStatus struct {
	ObservedGeneration int64  `json:"observedGeneration,omitempty"`
	Replicas           int32  `json:"replicas"`
	ReadyReplicas      int32  `json:"readyReplicas,omitempty"`
	CurrentReplicas    int32  `json:"currentReplicas,omitempty"`
	UpdatedReplicas    int32  `json:"updatedReplicas,omitempty"`
	AvailableReplicas  int32  `json:"availableReplicas,omitempty"`
	CurrentRevision    string `json:"currentRevision,omitempty"`
	UpdateRevision     string `json:"updateRevision,omitempty"`
	CollisionCount     *int32 `json:"collisionCount,omitempty"`
	Conditions         []WorkloadCondition `json:"conditions,omitempty"`
}

// WorkloadCondition 工作负载状态条件（通用）
type WorkloadCondition struct {
	Type               string `json:"type"`
	Status             string `json:"status"`
	Reason             string `json:"reason,omitempty"`
	Message            string `json:"message,omitempty"`
	LastUpdateTime     string `json:"lastUpdateTime,omitempty"`
	LastTransitionTime string `json:"lastTransitionTime,omitempty"`
}

// WorkloadRollout 工作负载发布状态
type WorkloadRollout struct {
	Phase   string   `json:"phase"`             // Progressing, Complete, Degraded, Paused
	Message string   `json:"message,omitempty"`
	Badges  []string `json:"badges,omitempty"`
}

// StatefulSet 辅助方法

func (s *StatefulSet) GetName() string {
	return s.Summary.Name
}

func (s *StatefulSet) GetNamespace() string {
	return s.Summary.Namespace
}

func (s *StatefulSet) IsHealthy() bool {
	return s.Summary.Ready == s.Summary.Replicas && s.Summary.Replicas > 0
}

func (s *StatefulSet) IsUpdating() bool {
	return s.Summary.Updated < s.Summary.Replicas
}

// ============================================================
// DaemonSet 模型（嵌套结构）
// ============================================================

// DaemonSet K8s DaemonSet 资源模型
type DaemonSet struct {
	Summary  DaemonSetSummary  `json:"summary"`
	Spec     DaemonSetSpec     `json:"spec"`
	Template PodTemplate       `json:"template"`
	Status   DaemonSetStatus   `json:"status"`
	Rollout  *WorkloadRollout  `json:"rollout,omitempty"`

	// 元数据
	Labels      map[string]string `json:"labels,omitempty"`
	Annotations map[string]string `json:"annotations,omitempty"`
}

// DaemonSetSummary DaemonSet 摘要
type DaemonSetSummary struct {
	Name                   string    `json:"name"`
	Namespace              string    `json:"namespace"`
	DesiredNumberScheduled int32     `json:"desiredNumberScheduled"`
	CurrentNumberScheduled int32     `json:"currentNumberScheduled"`
	NumberReady            int32     `json:"numberReady"`
	NumberAvailable        int32     `json:"numberAvailable"`
	NumberUnavailable      int32     `json:"numberUnavailable"`
	NumberMisscheduled     int32     `json:"numberMisscheduled"`
	UpdatedNumberScheduled int32     `json:"updatedNumberScheduled"`
	CreatedAt              time.Time `json:"createdAt"`
	Age                    string    `json:"age"`
	Selector               string    `json:"selector,omitempty"`
}

// DaemonSetSpec DaemonSet 规格
type DaemonSetSpec struct {
	UpdateStrategy       *UpdateStrategy `json:"updateStrategy,omitempty"`
	MinReadySeconds      int32           `json:"minReadySeconds,omitempty"`
	RevisionHistoryLimit *int32          `json:"revisionHistoryLimit,omitempty"`
	Selector             *LabelSelector  `json:"selector,omitempty"`
}

// DaemonSetStatus DaemonSet 状态
type DaemonSetStatus struct {
	ObservedGeneration     int64  `json:"observedGeneration,omitempty"`
	DesiredNumberScheduled int32  `json:"desiredNumberScheduled"`
	CurrentNumberScheduled int32  `json:"currentNumberScheduled"`
	NumberReady            int32  `json:"numberReady"`
	NumberAvailable        int32  `json:"numberAvailable,omitempty"`
	NumberUnavailable      int32  `json:"numberUnavailable,omitempty"`
	NumberMisscheduled     int32  `json:"numberMisscheduled"`
	UpdatedNumberScheduled int32  `json:"updatedNumberScheduled,omitempty"`
	CollisionCount         *int32 `json:"collisionCount,omitempty"`
	Conditions             []WorkloadCondition `json:"conditions,omitempty"`
}

// DaemonSet 辅助方法

func (d *DaemonSet) GetName() string {
	return d.Summary.Name
}

func (d *DaemonSet) GetNamespace() string {
	return d.Summary.Namespace
}

func (d *DaemonSet) IsHealthy() bool {
	return d.Summary.NumberReady == d.Summary.DesiredNumberScheduled && d.Summary.DesiredNumberScheduled > 0
}

func (d *DaemonSet) IsUpdating() bool {
	return d.Summary.UpdatedNumberScheduled < d.Summary.DesiredNumberScheduled
}

func (d *DaemonSet) HasMisscheduled() bool {
	return d.Summary.NumberMisscheduled > 0
}
